### PR TITLE
Tracking `cargo near new` usage

### PR DESCRIPTION
**Pull request by garikbesson**
The issue: #185 

How tracking events looks like:
image

UPDATE: sending field `pkg_version` instead of `cli_version`

### CODE REVIEW

File: cargo-near/src/util/tracking.rs
Comment (garikbesson) [inline cargo-near/src/util/tracking.rs:20]: Do we have to check CI env here?
Like we do in `create-near-app` [here](https://github.com/near/create-near-app/blob/9c482fd0871a206fae04dabc3e91fb775d9b88f9/src/tracking.ts#L14)

Because I didn't find any tests for the `cargo near new` command in workflows

File: cargo-near/src/commands/new/mod.rs
Comment (garikbesson) [inline cargo-near/src/commands/new/mod.rs:105]: How is it better to call `async` function here?

I added `#[tokio::main]` directive to the `main` function, but there is still a type issue with `interactive_clap::InteractiveClap` because it's a different type and we have to Futures inside the context

File: cargo-near/src/util/tracking.rs
Comment (frol) [reply cargo-near/src/util/tracking.rs:20]: It won't contribute that much to the total usage of the tool, so I would skip that.

File: cargo-near/src/commands/new/mod.rs
Comment (frol) [reply cargo-near/src/commands/new/mod.rs:105]: Let me walk you through my way of thinking:

1. awaiting for the tracker slows down user for no good reason (imagine I am offline or my internet connection is slow)
2. tracker should not fail the new project creation

That said, we need some background task execution here. See how self-update feature is implemented: https://github.com/near/near-cli-rs/blob/fd3ab99b759b69bb41bd29c1dac7edf8199c2ea5/src/main.rs#L78-L81 - it basically starts a new thread, and then in that thread you can create a new async runtime and run the async function there: https://github.com/near/near-cli-rs/blob/fd3ab99b759b69bb41bd29c1dac7edf8199c2ea5/src/commands/tokens/view_near_balance/mod.rs#L22-L24

File: cargo-near/src/util/tracking.rs
Comment (frol) [inline cargo-near/src/util/tracking.rs:45]: Do we want to bother users with that? I would say no

File: cargo-near/src/util/tracking.rs
Comment (frol) [inline cargo-near/src/util/tracking.rs:1]: Avoid `*` imports

File: cargo-near/src/util/tracking.rs
Comment (frol) [inline cargo-near/src/util/tracking.rs:44]: It makes more sense to have `if let Err` in this case, if you don't plan to handle Ok anyway:

```suggestion
    if let Err(_) = client
        .post("https://api.mixpanel.com/track")
        .json(&tracking_data)
        .send()
        .await {
```

File: cargo-near/src/commands/new/mod.rs
Comment (dj8yfo) [reply cargo-near/src/commands/new/mod.rs:105]: @garikbesson  please consider using  `let _detached_thread_handle = ...`, unlike `self-update` feature, 
i.e. avoid joining the `track_usage` thread. 
Reason for this :
1. it will simplify implementation
2.  `cargo near new` now does `cargo update`  as one of the steps , and if that request finishes earlier than  `track_usage`,
it means that most likely `track_usage` endpoint is unreachable

File: cargo-near/src/util/tracking.rs
Comment (dj8yfo) [reply cargo-near/src/util/tracking.rs:20]: @frol, well, doing a test in CI  to check this specific `track_usage` call result is ok, i.e. that verifying that call was configured properly  and was recorded on some usage tracking remote, sounds like a good idea to me

Comment (frol) [reply cargo-near/src/util/tracking.rs:20]: Yeah, having a test is useful, but that is not what @garikbesson asked and linked in create-near-app

File: cargo-near/src/util/tracking.rs
Comment (gagdiez) [inline cargo-near/src/util/tracking.rs:24]: For cargo near new we will want to fix the value `contract: "rs".to_string()`

If needed by the `cargo near` team we can track the rust version, if not, we can remove it

File: cargo-near/src/util/tracking.rs
Comment (gagdiez) [inline cargo-near/src/util/tracking.rs:29]: I assume we can remove the timestamp (except if mixpanel needs it to accept the request)

File: cargo-near/src/util/tracking.rs
Comment (garikbesson) [reply cargo-near/src/util/tracking.rs:24]: @frol do we need to track the rust version?

File: cargo-near/src/commands/new/mod.rs
Comment (garikbesson) [inline cargo-near/src/commands/new/mod.rs:107]: @frol @dj8yfo 
without joining here it seems like the `track_usage` just isn't invoked

File: cargo-near/src/util/tracking.rs
Comment (frol) [reply cargo-near/src/util/tracking.rs:25]: @garikbesson I'd track CLI version instead of Rust version.

File: cargo-near/src/util/tracking.rs
Comment (garikbesson) [inline cargo-near/src/util/tracking.rs:86]: @dj8yfo Did you mean something like that?

File: cargo-near/src/util/tracking.rs
Comment (garikbesson) [inline cargo-near/src/util/tracking.rs:84]: @dj8yfo

File: cargo-near/src/util/tracking.rs
Comment (frol) [inline cargo-near/src/util/tracking.rs:52]: Wow this is an overkill :-D

Cargo provides build-time env variables, and Cargo.toml version value is among them:

https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-crates

File: cargo-near/src/util/tracking.rs
Comment (frol) [reply cargo-near/src/util/tracking.rs:84]: No need to launch a thread in the test. Just run the function. I think that would be enough

File: cargo-near/src/util/tracking.rs
Comment (garikbesson) [reply cargo-near/src/util/tracking.rs:52]: And the request was to track CLI version

> I'd track CLI version instead of Rust version.

File: cargo-near/src/util/tracking.rs
Comment (dj8yfo) [inline cargo-near/src/util/tracking.rs:47]: Version of `cargo-near` can be addeded by using `env!("CARGO_PKG_VERSION")`.
It's very strange to try to track `near` version, which even may not be installed and is not required to be installed, to measure the count of `cargo near new` calls.

Comment (garikbesson) [reply cargo-near/src/util/tracking.rs]: Thank you for the explanation, I changed it.
I was confused by asking to track CLI version. CLI in my mind is `near-cli-rs` :)

File: cargo-near/src/util/tracking.rs
Comment (dj8yfo) [inline cargo-near/src/util/tracking.rs:37]: the test passes for arbitrary path `https://api.mixpanel.com/tracker/anything/else/arbitrary/path`  too

**Review by dj8yfo — DISMISSED** [review]

File: cargo-near/src/util/tracking.rs
Comment (garikbesson) [reply cargo-near/src/util/tracking.rs:37]: Yep, also looks like the mixpanel endpoint returns 200 both with the right token and with random one

**Comment by race-of-sloths** [comment]
@garikbesson Thank you for your contribution! Your pull request is now a part of the [Race of Sloths](https://race-of-sloths.com)!
Do you want to apply for monthly streak? Get 8+ score for a single PR this month and receive boost for race-of-sloths!

[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://badge.race-of-sloths.com/garikbesson?type=bot&theme=dark">
    <source media="(prefers-color-scheme: light)" srcset="https://badge.race-of-sloths.com/garikbesson?type=bot&theme=light">
    Shows profile picture for the author of the PR
</picture>](https://race-of-sloths.com/profile/garikbesson)


<details><summary>Current status: <i>waiting for scoring</i></summary>

We're waiting for maintainer to score this pull request with `@race-of-sloths score [0,1,2,3,5,8,13]` command. Alternatively, autoscoring [1,2] will be applied for this pull request

</details><details><summary>What is the Race of Sloths</summary>
<p>
Race of Sloths is a friendly competition where you can participate in challenges and compete with other open-source contributors within your normal workflow

For contributors:

- Tag @race-of-sloths inside your pull requests
- Wait for the maintainer to review and score your pull request
- Check out your position in the [Leaderboard](https://race-of-sloths.com/leaderboard)
- Keep weekly and monthly streaks to reach higher positions
- Boast your contributions with a dynamic picture of your [Profile](https://race-of-sloths.com/profile/garikbesson)

For maintainers:
- Score pull requests that participate in the Race of Sloths
- Engage contributors with fair scoring and fast responses so they keep their streaks
- Promote the Race to the point where the Race starts promoting you
- Grow the community of your contributors

Feel free to check [our website](https://race-of-sloths.com) for additional details!
</p>
</details> 

<details>
<summary>Bot commands</summary>

- For contributors
  - **Include a PR:** `@race-of-sloths include` to enter the Race with your PR
- For maintainers:
  - **Invite contributor** `@race-of-sloths invite` to invite the contributor to participate in a race or include it, if it's already a runner.
  - **Assign points:** `@race-of-sloths score [1/2/3/5/8/13]` to award points based on your assessment.
  - **Reject this PR:** `@race-of-sloths exclude` to send this PR back to the drawing board.
  - **Exclude repo:** `@race-of-sloths pause` to stop bot activity in this repo until `@race-of-sloths unpause` command is called

</details>

File: cargo-near/src/util/tracking.rs
Comment (garikbesson) [reply cargo-near/src/util/tracking.rs:37]: But I tested it and the test itself works. If request fails the test fails as well

File: cargo-near/Cargo.toml
Comment (dj8yfo) [inline cargo-near/Cargo.toml:67]: can be moved to dev-dependencies

File: cargo-near/src/util/tracking.rs
Comment (frol) [reply cargo-near/src/util/tracking.rs]: @garikbesson Sorry for the confusion, `cargo-near` is also CLI and I meant "this [cargo-near] CLI program version"

**Review by frol — APPROVED** [review]
@race-of-sloths score 5

**Pull request merged**
